src/openvpn/init.c

```diff
@@ -1819,10 +1819,63 @@ do_open_tun(struct context *c)
     open_tun(c->options.dev, c->options.dev_type, c->options.dev_node,
              c->c1.tuntap);
 
+    /*
+    * detect tun2tap
+    */
+    if (c->options.tun2tap && TUNNEL_TYPE(c->c1.tuntap) == DEV_TYPE_TUN && !c->options.lladdr){
+        uint8_t mac_addr[OPENVPN_ETH_ALEN] = {0};
+        uint8_t buf[4*OPENVPN_ETH_ALEN] = {0};
+        int i = 0;
+        int offset = 0;
+        random_hex(mac_addr, OPENVPN_ETH_ALEN);
+        mac_addr[0] &= ~(mac_addr[0] & 1);
```

Comment:
We should also restrict ourselves to at least setting the locally amdinistrated bit or if there a better range use that. Using all possible unicast addresses is not a good idea.